### ISSUE TITLE
[RFC] SchemaStore sync

### ISSUE THREAD

**Issue by guilliamxavier** [composer/composer#9903]
Greetings,

After #9782/#9897 and https://github.com/SchemaStore/schemastore/pull/1617, the remaining diff between https://github.com/composer/composer/blob/master/res/composer-schema.json and https://github.com/SchemaStore/schemastore/blob/master/src/schemas/json/composer.json is:

```diff
@@ -61,7 +61,7 @@
         "version": {
             "type": "string",
             "description": "Package version, see https://getcomposer.org/doc/04-schema.md#version for more info on valid schemes.",
-            "pattern": "^v?\\d+(\\.\\d+){0,3}|^dev-"
+            "pattern": "^v?\\d+(\\.\\d+){0,3}(-(dev|(p(atch)?|a(lpha)?|b(eta)?|RC)(\\d+)?))?$"
         },
         "time": {
             "type": "string",
```

- The SchemaStore-modified pattern is very close to https://getcomposer.org/doc/04-schema.md#version

  > This must follow the format of `X.Y.Z` or `vX.Y.Z` with an optional suffix of `-dev`, `-patch` (`-p`), `-alpha` (`-a`), `-beta` (`-b`) or `-RC`. The patch, alpha, beta and RC suffixes can also be followed by a number.

  but actually also allows `X`, `X.Y` and `X.Y.Z.W` in addition to `X.Y.Z` (like the "official" pattern).
- The "official" pattern only tests the start (see #8262 for initial discussion), so also allows metadata like in "1.2.3+foo" (or "1.2.3-p1+bar") but also invalid things like "1.foo bar!"; and additionally allows "dev-_anything_" (which again includes invalid things).

Moreover, shortly after, this was added here: https://github.com/composer/composer/blob/8a3f4a84000485734b6e3e269480a8a1a8ab55a8/res/composer-schema.json#L182-L188 but has yet to be repeated on SchemaStore (manually).

See also https://github.com/composer/composer/pull/9782#issuecomment-801242320 (the good news is that the official schema is valid for SchemaStore now!).

So:

1. Can we reach a "consensusual" version pattern?
2. To avoid the need of manual sync after each change to the official schema, can https://github.com/SchemaStore/schemastore/commit/59bab3ed8397440f8bf45c9c7b16f3df21ea1eea (Gruntfile) be reverted (@madskristensen)?

Thanks 🙂

**Comment by Seldaek** [comment]
1. The version pattern `^v?\\d+(\\.\\d+){0,3}(-(dev|(p(atch)?|a(lpha)?|b(eta)?|RC)(\\d+)?))?$` is definitely incomplete for composer, as it ignores dev versions. Versions are very rarely defined in composer.json directly, so validating more than we do in our schema is IMO not strictly needed. See https://github.com/composer/semver/blob/main/tests/VersionParserTest.php#L57 if you really want to make a regex that works for all that we support though, but IMO it's gonna be a giant mess and error prone and we will probably miss valid use cases, so I'd rather not extend it to be fully compliant.
2. I guess it can be reverted now.. but that's not up to me :)

So in short I'd do 2, and forget about 1 (which wouldn't be a problem anymore if 2 just syncs up our schema the diff goes away right?).

**Comment by guilliamxavier** [comment]
Thanks, closing in favor of https://github.com/SchemaStore/schemastore/pull/1618
